### PR TITLE
feat: wire live context token counts through engine to compaction guards

### PR DESCRIPTION
**Pull request by 100yenadmin**
## Summary

Wires live observed token counts from the engine layer into the compaction guards added in #306, so headroom and cache-aware skip decisions use fresh data instead of potentially stale stored counts. Also propagates the new config fields (`leafSkipReductionThreshold`, `leafBudgetHeadroomFactor`) from `LcmConfig` into `CompactionConfig`.

**Part 2 of 3 from #289 split. Depends on #306. Merge order: #306 → #307 → #308.**

---

## The Problem: Stale Token Counts Cause Wrong Guard Decisions

The compaction guards in #306 decide whether to skip or compact based on `totalAssembledTokens` — a value derived from the stored token count in the database. But stored counts can lag behind reality:

### Stale-low scenario (most dangerous)

After rapid message ingestion (e.g., a tool that emits 15 messages in 1 second), the DB count hasn't caught up:
- **Stored count**: 30K (stale from 2 seconds ago)
- **Live count**: 75K (actual prompt tokens the API will see)
- **Budget ceiling**: 60K

Without live counts, the headroom guard sees 30K < 60K → **SKIP**. But the context is actually 75K — well over the ceiling. The guard should detect budget pressure and **COMPACT**.

### The fix: `max(stored, live)`

By passing `liveContextTokens` (from `estimateSessionTokenCountForAfterTurn`) through to `evaluateLeafTrigger`, the guard uses whichever count is higher:
- Stale-low stored → uses live (prevents missed compaction)
- Stale-high stored → uses stored (conservative, prevents premature skip)

This is the safe/conservative choice — it errs on the side of compacting when counts disagree, which is the correct bias for preventing context overflow.

---

### Engine → CompactionEngine config

`LcmConfig` now flows `leafSkipReductionThreshold` and `leafBudgetHeadroomFactor` into the `CompactionConfig` object, so plugin/env var overrides actually take effect at runtime (without this, the guards always use their internal defaults).

### afterTurn → evaluateLeafTrigger

The engine's `afterTurn()` already computes `liveContextTokens` — this PR passes it (along with `tokenBudget`) through to `evaluateLeafTrigger()` so the guards can make informed decisions.

### afterTurn logging

New structured logging for compaction decisions:
- **Trigger**: `[lcm] afterTurn: leaf compaction triggered (raw=24000, threshold=20000, assembled=548000, pressure=true)`
- **Skip**: `[lcm] afterTurn: leaf compaction skipped (budget-headroom: 30000 assembled < 120000 ceiling)`

These match the log lines documented in the tuning guide (#308).

### compactLeaf/compactFullSweep → currentTokenCount

Both paths now pass `observedTokens` (from `normalizeObservedTokenCount`) as `currentTokenCount`, which `evaluateLeafTrigger` uses as `precomputedTokenCount` to avoid a duplicate DB read.

### compact() type fix

`compact()` wrapper now includes `currentTokenCount` in its input type so TypeScript excess-property checks pass and the value flows through to `compactFullSweep`.

### LeafTriggerResult import

Engine's `evaluateLeafTrigger` now imports `LeafTriggerResult` from `compaction.ts` instead of re-declaring the shape inline, preventing type drift.

### README env var table

Added `LCM_LEAF_SKIP_REDUCTION_THRESHOLD`, `LCM_LEAF_BUDGET_HEADROOM_FACTOR`, and `LCM_FALLBACK_PROVIDERS` to the README environment variable reference table.

---

## Changes by File

| File | Lines | Change |
|------|-------|--------|
| `src/engine.ts` | +26/-3 | Pass config fields to CompactionConfig. Pass `tokenBudget` + `liveContextTokens` to `evaluateLeafTrigger` from afterTurn. Pass `currentTokenCount` to `compactLeaf`/`compact`. Import `LeafTriggerResult`. Add trigger/skip logging. |
| `src/compaction.ts` | +1 | Add `currentTokenCount` to `compact()` input type. |
| `test/engine.test.ts` | +48/-1 | Update `evaluateLeafTrigger` assertion for 4-arg signature. Add `currentTokenCount` to compact plumbing assertions. New test: `compactLeafAsync` passes `currentTokenCount`. New test: omission when not provided. |
| `test/lcm-integration.test.ts` | +2 | Shrink test message bodies (was 12KB strings, now short descriptive text). |
| `README.md` | +3 | Add 3 env vars to reference table. |

## Test Plan

- [x] 204 tests passing (122 engine + 82 integration)
- [x] `evaluateLeafTrigger` called with `(sessionId, sessionKey, tokenBudget, liveContextTokens)`
- [x] `currentTokenCount: 500` flows through compact plumbing to `compactFullSweep`
- [x] `currentTokenCount` omitted from `compactLeaf` call when not provided (no undefined leakage)
- [x] Stale-token integration tests: compactLeaf and compactFullSweep trigger with live counts

### CODE REVIEW

File: src/engine.ts [hunk 1]
```diff
@@ -3103,6 +3122,7 @@ export class LcmContextEngine implements ContextEngine {
           const sweepResult = await this.compaction.compact({
             conversationId,
             tokenBudget,
+            currentTokenCount: observedTokens,
             summarize,
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply src/engine.ts:3113]: Good catch — fixed in 60b4a31. Added `currentTokenCount?: number` to `compact()`'s input type so TS excess-property checks pass and the value flows through to `compactFullSweep()`.

File: src/engine.ts [hunk 2]
```diff
@@ -2837,10 +2846,19 @@ export class LcmContextEngine implements ContextEngine {
...
+    skipReason?: string;
+    context?: {
+      totalAssembledTokens: number;
+      budgetCeiling?: number;
+      budgetPressure: boolean;
+      estimatedReduction?: number;
+      reductionThreshold?: number;
+      headroomFactor: number;
+    };
   }> {
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply src/engine.ts]: Agreed — fixed in 60b4a31. Now imports `LeafTriggerResult` from `./compaction.js` instead of re-declaring the shape inline. Eliminates type drift.

File: .changeset/cache-aware-compaction-guards.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+"@martian-engineering/lossless-claw": minor
+---
+
+Cache-aware leaf compaction guards with budget-pressure override and per-tier tuning. Prevents unnecessary prompt-cache invalidation by skipping compaction when token reduction is negligible or budget headroom is ample.
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply .changeset/cache-aware-compaction-guards.md]: This changeset file is inherited from the base branch (`feat/cache-aware-compaction-guards`, PR #306) — it's not a new addition by this PR. Since #306 merges first with this changeset, it will already be on main when #307 merges. The frontmatter was also fixed in #306 (commit 9425542).

File: src/compaction.ts [hunk 4]
```diff
@@ -436,17 +456,95 @@ export class CompactionEngine {
...
+    const storedTokens = precomputedTokenCount ?? await this.summaryStore.getContextTokenCount(conversationId);
+    const normalizedLiveContextTokens =
+      Number.isFinite(liveContextTokens) && (liveContextTokens as number) > 0
+        ? Math.floor(liveContextTokens as number)
+        : undefined;
+    const totalAssembledTokens =
+      normalizedLiveContextTokens !== undefined && normalizedLiveContextTokens > storedTokens
+        ? normalizedLiveContextTokens
+        : storedTokens;
+
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply src/compaction.ts:492]: This is by design. The `max(stored, live)` logic handles the primary problem case: stored counts are stale-**low** (e.g., after rapid ingestion outpaces DB writes), causing the guards to incorrectly skip compaction. When stored counts are stale-**high** (e.g., after externalization), using the higher value is the safe/conservative choice — it prevents premature compaction skipping and lets the guards err on the side of compacting rather than not.

File: src/db/config.ts [hunk 5]
```diff
@@ -281,5 +287,17 @@ export function resolveLcmConfig(
...
+      parseFiniteNumber(env.LCM_LEAF_SKIP_REDUCTION_THRESHOLD)
+        ?? toNumber(pc.leafSkipReductionThreshold) ?? 0.05,
+    ),
+    leafBudgetHeadroomFactor: clamp01(
+      parseFiniteNumber(env.LCM_LEAF_BUDGET_HEADROOM_FACTOR)
+        ?? toNumber(pc.leafBudgetHeadroomFactor) ?? 0.8,
+    ),
   };
 }
+
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply src/db/config.ts:300]: Good catch — fixed in 8ea347f. Added `LCM_LEAF_SKIP_REDUCTION_THRESHOLD`, `LCM_LEAF_BUDGET_HEADROOM_FACTOR`, and `LCM_FALLBACK_PROVIDERS` to the README env var table.

File: test/lcm-integration.test.ts [hunk 6]
```diff
@@ -2126,6 +2126,54 @@ describe("LCM integration: compaction", () => {
...
+  it("compactLeaf uses currentTokenCount for headroom decisions when stored tokens are stale", async () => {
+    const staleAwareEngine = new CompactionEngine(convStore as any, sumStore as any, {
+      ...defaultCompactionConfig,
+      leafChunkTokens: 15_000,
+    });
+
+    await ingestMessages(convStore, sumStore, 10, {
+      contentFn: (i) => `Turn ${i}: ${"x".repeat(11_980)}`,
+      tokenCountFn: () => 3_000,
+    });
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply test/lcm-integration.test.ts:2138]: Good point — fixed in 8ea347f. Replaced the 12KB `"x".repeat(11_980)` strings with short descriptive text since `tokenCountFn` overrides the count anyway.

File: README.md [hunk 7]
```diff
@@ -157,6 +157,9 @@ Add a `lossless-claw` entry under `plugins.entries` in your OpenClaw config:
 | `LCM_DELEGATION_TIMEOUT_MS` | `120000` | Max time to wait for delegated `lcm_expand_query` sub-agent completion |
 | `LCM_SUMMARY_TIMEOUT_MS` | `60000` | Max time to wait for a single model-backed LCM summarizer call |
 | `LCM_PRUNE_HEARTBEAT_OK` | `false` | Retroactively delete `HEARTBEAT_OK` turn cycles from LCM storage |
+| `LCM_LEAF_SKIP_REDUCTION_THRESHOLD` | `0.05` | Minimum estimated reduction (fraction of assembled tokens) to justify leaf compaction; set to 0 to disable cache-aware skip |
+| `LCM_LEAF_BUDGET_HEADROOM_FACTOR` | `0.8` | Skip leaf compaction when assembled tokens are below this fraction of the budget ceiling; set to 0 to disable headroom check |
+| `LCM_FALLBACK_PROVIDERS` | `""` | Comma-separated `provider:model` pairs for compaction summarization fallbacks |
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply README.md]: Good catch — fixed in 87c7df2. Changed to `provider/model` with an example (`anthropic/claude-haiku-4-5,openai/gpt-4o-mini`) to match the parser's `/` delimiter.

File: src/compaction.ts [hunk 8]
```diff
@@ -436,17 +456,95 @@ export class CompactionEngine {
...
+    if (rawTokensOutsideTail < threshold) {
+      return { shouldCompact: false, rawTokensOutsideTail, threshold };
+    }
+
+    // Reuse a precomputed token count when the caller already fetched it
+    // (avoids a duplicate getContextTokenCount DB read).
+    // Use the higher of stored/precomputed count and live context estimate
+    // for more accurate headroom decisions.
+    const storedTokens = precomputedTokenCount ?? await this.summaryStore.getContextTokenCount(conversationId);
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply src/compaction.ts:483]: Fixed in 9b9a04b. Added a short-circuit before the DB read: when both `leafBudgetHeadroomFactor <= 0` and `leafSkipReductionThreshold <= 0`, returns `shouldCompact: true` immediately without fetching `storedTokens`.

File: src/db/config.ts [hunk 9]
```diff
@@ -281,5 +287,17 @@ export function resolveLcmConfig(
...
       parseFallbackProviders(env.LCM_FALLBACK_PROVIDERS)
         ?? toFallbackProviderArray(pc.fallbackProviders) ?? [],
+    leafSkipReductionThreshold: clamp01(
+      parseFiniteNumber(env.LCM_LEAF_SKIP_REDUCTION_THRESHOLD)
+        ?? toNumber(pc.leafSkipReductionThreshold) ?? 0.05,
+    ),
+    leafBudgetHeadroomFactor: clamp01(
+      parseFiniteNumber(env.LCM_LEAF_BUDGET_HEADROOM_FACTOR)
+        ?? toNumber(pc.leafBudgetHeadroomFactor) ?? 0.8,
+    ),
```
[the remark this answers is not in the document]
Comment (100yenadmin) [reply src/db/config.ts:297]: Fixed in 9b9a04b. Added boundary-value tests: `leafSkipReductionThreshold: -0.5` clamps to 0, `leafBudgetHeadroomFactor: 2.0` clamps to 1.

**Comment by 100yenadmin** [comment]
This branch looks functionally superseded now. The runtime behavior it was pushing toward landed through #318 and #329, and the branch itself is now conflicting with current `main`.

I'm treating this as replaceable history rather than something worth reviving directly. The live follow-up work from the cost/compaction sweep will build on current `main`, not on this branch.

**Comment by 100yenadmin** [comment]
Closing as superseded by merged runtime work on current `main`, primarily #318 and #329. Any remaining cost/compaction follow-up will be rebased onto current `main` rather than carried on this conflicting branch.
